File: src/ai_notify/helpers/filters.py
"""
Notification filtering logic.
"""

from loguru import logger
from ai_notify.config_loader import AINotifyConfig, NotificationMode


def should_send_notification(
    prompt: str,
    duration_seconds: int,
    config: AINotifyConfig,
) -> bool:
    """
    Determine whether to send a notification based on filters.

    Args:
        prompt: User prompt text
        duration_seconds: Job duration in seconds
        config: Runtime configuration

    Returns:
        True if notification should be sent, False otherwise
    """
    # Check duration threshold
    threshold = config.notification.threshold_seconds
    if duration_seconds < threshold:
        logger.debug(
            f"Skipping notification: duration {duration_seconds}s < threshold {threshold}s"
        )
        return False

    # Check exclude patterns (case-sensitive prefix matching)
    exclude_patterns = config.notification.exclude_patterns
    if exclude_patterns and prompt:
        for pattern in exclude_patterns:
            if prompt.startswith(pattern):
                logger.debug(
                    f"Skipping notification: prompt starts with excluded pattern '{pattern}'"
                )
                return False

    return True


def should_send_completion_notification(
    prompt: str,
    duration_seconds: int,
    config: AINotifyConfig,
) -> bool:
    """
    Determine whether to send a completion notification based on mode and filters.

    Args:
        prompt: User prompt text
        duration_seconds: Job duration in seconds
        config: Runtime configuration

    Returns:
        True if notification should be sent, False otherwise
    """
    # Check mode first (fast path)
    if config.notification.mode == NotificationMode.DISABLED:
        return False

    if config.notification.mode == NotificationMode.PERMISSION_ONLY:
        return False

    # Mode is ALL - apply existing filters
    return should_send_notification(prompt, duration_seconds, config)


def should_send_permission_notification(config: AINotifyConfig) -> bool:
    """
    Determine whether to send permission request notifications.

    Args:
        config: Runtime configuration

    Returns:
        True if permission notifications should be sent, False otherwise
    """
    return config.notification.mode != NotificationMode.DISABLED


def should_send_codex_notification(prompt: str, config: AINotifyConfig) -> bool:
    """
    Determine whether to send a Codex CLI notification.

    Codex notify payloads do not include job duration, so we only apply mode
    and exclude pattern filtering.

    Args:
        prompt: User prompt text
        config: Runtime configuration

    Returns:
        True if notification should be sent, False otherwise
    """
    if config.notification.mode == NotificationMode.DISABLED:
        return False

    if config.notification.mode == NotificationMode.PERMISSION_ONLY:
        return False

    exclude_patterns = config.notification.exclude_patterns
    if exclude_patterns and prompt:
        for pattern in exclude_patterns:
            if prompt.startswith(pattern):
                logger.debug(
                    f"Skipping Codex notification: prompt starts with excluded pattern '{pattern}'"
                )
                return False

    return True

File: src/filters.rs
//! Pure notification filtering rules.

use crate::{config::AppConfig, model::NotificationMode};

/// Applies the duration and case-sensitive prompt-prefix filters.
pub fn should_send_notification(prompt: &str, duration_seconds: u64, config: &AppConfig) -> bool {
    duration_seconds >= config.notification.threshold_seconds &&
        !config.notification.exclude_patterns.iter().any(|pattern| !prompt.is_empty() && prompt.starts_with(pattern))
}

/// Completion notifications are only enabled in `all` mode.
pub fn should_send_completion_notification(prompt: &str, duration_seconds: u64, config: &AppConfig) -> bool {
    config.notification.mode == NotificationMode::All && should_send_notification(prompt, duration_seconds, config)
}

/// Permission and question notifications are enabled unless all notifications are disabled.
pub fn should_send_permission_notification(config: &AppConfig) -> bool {
    config.notification.mode != NotificationMode::Disabled
}

/// API failures intentionally bypass duration and prompt filters.
pub fn should_send_failure_notification(config: &AppConfig) -> bool {
    config.notification.mode == NotificationMode::All
}

/// Codex payloads have no duration, so only mode and prompt-prefix filtering applies.
pub fn should_send_codex_notification(prompt: &str, config: &AppConfig) -> bool {
    config.notification.mode == NotificationMode::All &&
        !config.notification.exclude_patterns.iter().any(|pattern| !prompt.is_empty() && prompt.starts_with(pattern))
}

#[cfg(test)]
mod tests {
    use crate::{config::NotificationConfig, model::NotificationMode};

    use super::*;

    fn config() -> AppConfig {
        AppConfig {
            notification: NotificationConfig {
                threshold_seconds: 10,
                exclude_patterns: vec!["/skip".into()],
                ..NotificationConfig::default()
            },
            ..AppConfig::default()
        }
    }

    #[test]
    fn duration_and_case_sensitive_prefixes_filter_completions() {
        let config = config();
        assert!(!should_send_notification("work", 9, &config));
        assert!(!should_send_notification("/skip work", 10, &config));
        assert!(should_send_notification("/Skip work", 10, &config));
        assert!(should_send_notification("work /skip", 10, &config));
    }

    #[test]
    fn notification_mode_matrix_is_preserved() {
        let mut config = config();
        for mode in [NotificationMode::PermissionOnly, NotificationMode::Disabled] {
            config.notification.mode = mode;
            assert!(!should_send_completion_notification("work", 99, &config));
            assert!(!should_send_codex_notification("work", &config));
        }
        config.notification.mode = NotificationMode::PermissionOnly;
        assert!(should_send_permission_notification(&config));
        assert!(!should_send_failure_notification(&config));
        config.notification.mode = NotificationMode::Disabled;
        assert!(!should_send_permission_notification(&config));
        config.notification.mode = NotificationMode::All;
        assert!(should_send_failure_notification(&config));
    }

    #[test]
    fn failure_ignores_regular_filters() {
        let config = config();
        assert!(should_send_failure_notification(&config));
        assert!(!should_send_completion_notification("/skip", 1, &config));
    }
}
